manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,729 @@
+		return
+	}
+
+	// we know, because we allocated this network, that its IPAM field is fully
+	// allocated, and we can use it without nil checking
+	ipam, _ := a.drvRegistry.IPAM(local.nw.IPAM.Driver.Name)
+
+	for _, config := range local.nw.IPAM.Configs {
+		// these things can return errors, but we literally can't do anything
+		// about it if they do, so just ignore it.
+		ipam.ReleaseAddress(local.pools[config.Subnet], net.ParseIP(config.Gateway))
+		ipam.ReleasePool(local.pools[config.Subnet])
+	}
+	delete(a.networks, network.ID)
+}
+
+// AllocateVIPs allocates the VIPs for the provided endpoint and network ids.
+func (a *allocator) AllocateVIPs(endpoint *api.Endpoint, networkIDs []string) (rerr error) {
+	// if the endpoint spec mode has changed to DNSRR, then what we're actually
+	// doing is freeing all of the endpoint specs.
+	if endpoint.Spec != nil {
+		if endpoint.Spec.Mode == api.ResolutionModeDNSRoundRobin {
+			a.deallocateVIPs(endpoint.VirtualIPs)
+			return nil
+		}
+	}
+	// first, go through and check that every network we want a VIP for is
+	// allocated. if not, return an error. We can't allocate VIPs until the
+	// network is allocated
+	for _, nwid := range networkIDs {
+		if _, ok := a.networks[nwid]; !ok {
+			return ErrNetworkNotAllocated{nwid}
+		}
+	}
+
+	// now compute the changes
+	allocate := []string{}
+	// keep is the set of all virtual IPs we'll retain between the previous and
+	// current spec. we make it with a capacity the same as the current vips
+	// because typically, the vips in use won't change much, and we can avoid
+	// allocation by using this as a guess
+	keep := make([]*api.Endpoint_VirtualIP, 0, len(endpoint.VirtualIPs))
+	deallocate := []*api.Endpoint_VirtualIP{}
+	// first, we need to figure out if any virtual IPs are being removed
+	// continues are bad and hard to follow, so here's the plain english
+	// explanation:
+	// for every VIP currently allocated
+	//     go through the list of desired network IDs
+	//         if a network ID matches the ID on the VIP
+	//             then we're keeping this vip, so add it to the keep list and
+	//             go to the next VIPs, skipping the bottom of this loop
+	//     if we get to this point, then we have been through every desired
+	//     network ID and not found one matching the one on this VIP, so we
+	//     can add it to the list of VIPs to deallocate
+vips:
+	for _, vip := range endpoint.VirtualIPs {
+		for _, nwid := range networkIDs {
+			if vip.NetworkID == nwid {
+				keep = append(keep, vip)
+				continue vips
+			}
+		}
+		deallocate = append(deallocate, vip)
+	}
+
+	// now figure out which new network IDs we've added, which is the same loop
+	// above but swapped around to check network IDs against VIPs
+newvips:
+	for _, nwid := range networkIDs {
+		for _, vip := range endpoint.VirtualIPs {
+			if vip.NetworkID == nwid {
+				continue newvips
+			}
+		}
+		allocate = append(allocate, nwid)
+	}
```

Comment:
I'd prefer to avoid code with `continue LABEL` if possible. These two loops could written to loop over a `map[string]struct{}` which acts as a set of the requested network IDs. Likewise, we could have a mapping of existing vip network IDs to the VIP. This would allow us to replace these 2 doubly-nested loops with a single loop:

```go
for netID, currentVIP := range currentVIPsByNetID {
	if _, shouldKeep := requestedNetIDs[netID]; shouldKeep {
		keep = append(keep, currentVIP)
	} else {
		deallocate = append(deallocate, currentVIP)
	}
	delete(requestedNetIDs, netID)
}

// Now `keep` contains all of the VIPs we should keep, `deallocate` contains
// all of the VIPs we should deallocate, and any network IDs which remain in
// the `requestedNetIDs` set are IDs for networks from which we need to
// allocate new VIPs.
```

Reply:
yes you are correct and i agree.